execution/block-partitioner/src/pre_partition/connected_component/mod.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    pre_partition::PrePartitioner,
+    v2::{
+        load_balance::longest_processing_time_first,
+        state::PartitionState,
+        types::{TxnIdx0, TxnIdx1},
+        union_find::UnionFind,
+    },
+};
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+/// A `PrePartitioner` used in `PartitionerV2` that tries to group conflicting txns together (but a group size limit applied),
+/// then assign the groups to the shards using Longest-processing-time-first (LPT) scheduling.
+/// https://en.wikipedia.org/wiki/Longest-processing-time-first_scheduling
+///
+/// Note that the number of groups varies depending on the block:
+/// - if the conflict level is high and group size limit is loose, we may have only 1 group;
+/// - if the all txns are independent, each will be in its own group.
+///
+/// The group size limit is controlled by parameter `load_imbalance_tolerance` in the following way:
+/// if `block_size=100, num_shards=10, load_imbalance_tolerance=2.0`,
+/// then the size of a conflicting txn group is not allowed to exceed 100/10*2.0 = 20.
+/// This fact, combined with the LPT algorithm, guarantees that shard load will not exceed 20.
+pub struct ConnectedComponentPartitioner {
+    pub load_imbalance_tolerance: f32,
+}
+
+impl PrePartitioner for ConnectedComponentPartitioner {
+    fn pre_partition(&self, state: &mut PartitionState) {
```

Comment:
This is currently insignificant compared with the other parts.
With 32 threads, the overall partition TPS is still 200-400k depending on the conflict pattern.

In case we want to parallelize it, there are plenty distributed union-find algorithms we can explore.